spark3/src/main/java/org/apache/spark/sql/connector/read/SupportsReadFileFilter.java

```diff
@@ -0,0 +1,26 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.spark.sql.connector.read;
+
+import java.util.Set;
+
+public interface SupportsReadFileFilter extends Scan {
+  void filterFiles(Set<String> locations);
```

Comment:
The MERGE INTO doc is more generic than files and thinks of each records as part of a "group" that is deleted at once. That group could be a file or a partition.

We could use that to make this generic by renaming it to `filterGroups` and passing in an `Iterator` of group identifiers. Then we would need to have a way to project the group identifier using a metadata column.

This probably isn't something we're interested in doing just for Iceberg, but it is a change we would likely make to get this upstream.